neverbleed.h

```diff
@@ -41,14 +41,34 @@ typedef struct st_neverbleed_t {
     unsigned char auth_token[NEVERBLEED_AUTH_TOKEN_SIZE];
 } neverbleed_t;
 
+#define ROUND2WORD(n) (n + 64 - 1 - (n + 64 - 1) % 64)
+#define BITMASK(b) (1 << ((b) % CHAR_BIT))
+#define BITBYTE(b) ((b) / CHAR_BIT)
+#define BITSET(a, b) ((a)[BITBYTE(b)] |= BITMASK(b))
+#define BITUNSET(a, b) ((a)[BITBYTE(b)] &= ~BITMASK(b))
+#define BITBYTES(nb) ((nb + CHAR_BIT - 1) / CHAR_BIT)
+#define BITCHECK(a, b) ((a)[BITBYTE(b)] & BITMASK(b))
+
 /**
  * initializes the privilege separation engine (returns 0 if successful)
  */
 int neverbleed_init(neverbleed_t *nb, char *errbuf);
 /**
  * loads a private key file (returns 1 if successful)
  */
-int neverbleed_load_private_key_file(neverbleed_t *nb, SSL_CTX *ctx, const char *fn, char *errbuf);
+int neverbleed_load_private_key_file_index(neverbleed_t *nb, SSL_CTX *ctx, const char *fn, char *errbuf, size_t *key_index, size_t *type);
+static inline int neverbleed_load_private_key_file(neverbleed_t *nb, SSL_CTX *ctx, const char *fn, char *errbuf) {
+    return neverbleed_load_private_key_file_index(nb, ctx, fn, errbuf, NULL, NULL);
+}
+
+/*
+ * deletes a private key at key index position (returns 1 if successful)
+ */
+int neverbleed_del_rsa_key(neverbleed_t *nb, const size_t key_index);
+#if (!defined(LIBRESSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER >= 0x1010000fL)
+int neverbleed_del_ecdsa_key(neverbleed_t *nb, const size_t key_index);
```

Comment:
Am I correct in assuming that the expected pattern of using neverbleed proposed in this PR is:
* load certificate using `neverbleed_load_private_key_file_index`
* unload certificate using `neverbleed_del_???_key` and also calling `SSL_free` (or `EVP_PKEY_free`)

If that is the case, I am afraid if we can guarantee that all the use of the keys can precede the call to `neverbleed_del_???_key`.

How about setting the finish callbacks of the proxy objects so that the "del_rsa_key" or "del_ecdsa_key" IPC message can be sent at the exact moment when OpenSSL destroys the proxy objects? I'd assume that doing so will simplify the API and lower the cost of maintaining the code-base.
FWIW, I think that the finish callbacks can be set by calling `RSA_meth_set_finish` / `EC_KEY_METHOD_set_init`.

I am sorry for not asking this earlier, but I will appreciate it if you could consider the approach.

Reply:
FWIW, test.c on [kazuho/test.c-with-hup branch](https://github.com/h2o/neverbleed/tree/kazuho/test.c-with-hup) implements a test server that reloads the keys and certificates when receiving a SIGHUP. I think that we can use it for testing the PR for checking leaks and / or errors under stress.

Reply:
Yes that's the usage we had in mind and also the concern (about using after freeing).
thank you for the suggestion! Will definitely look into that and would like very much to simplify the API.

Thanks for the stress tester - will definitely use it. The one I've been using exercises the particular use-case where partial removals leave gaps, which are supposed to be found and reused by the load key calls.